src/controller/CHIPDeviceController.cpp

```diff
@@ -812,7 +812,27 @@ void DeviceCommissioner::RendezvousCleanup(CHIP_ERROR status)
 
 void DeviceCommissioner::OnSessionEstablishmentError(CHIP_ERROR err)
 {
-    // PASE session establishment failure.
+    // Need to null out mDeviceInPASEEstablishment before maybe trying to
+    // establish PASE again, so EstablishPASEConnection won't error out.
+    auto * oldDeviceInPASEEstablishment = mDeviceInPASEEstablishment;
+    mDeviceInPASEEstablishment          = nullptr;
+
+    if (mSetUpCodePairer.TryNextRendezvousParameters())
+    {
+        ChipLogProgress(Controller, "Ignoring PASE error; will try commissioning over a different transport");
+        // We don't need oldDeviceInPASEEstablishment anymore.
+        ReleaseCommissioneeDevice(oldDeviceInPASEEstablishment);
+        return;
+    }
+
+    // Reset mDeviceInPASEEstablishment because PairingFailed checks for that to
+    // send the right notifications.
+    mDeviceInPASEEstablishment = oldDeviceInPASEEstablishment;
+    PairingFailed(err);
+}
+
+void DeviceCommissioner::PairingFailed(CHIP_ERROR err)
```

Comment:
What is the difference between "Pairing" and "Session Establishment"? This is coupling the SetupCodePairer logic to an extent to the CHIPDeviceController...

There is really non-trivial state logic here.

Reply:
> What is the difference between "Pairing" and "Session Establishment"?

Fair question.  "Pairing" in this case means "all attempts to establish a session have failed and we have nothing else to try".  I am open to better suggestions.

> This is coupling the SetupCodePairer logic to an extent to the CHIPDeviceController...

They're already coupled pretty tightly...  I agree that ideally they would be decoupled entirely, but that is  much bigger project with a lot of risk of causing unintended problems, so seems like post-1.0 work at this point.

> There is really non-trivial state logic here.

Yes, including before this change...

Reply:
The coupling is very directional: CHIPDeviceController doesn't rely on SetupCodePairer, really (we use it without SetupCodePairer usage), but the opposite is not true.

Reply:
That doesn't change here, right?  There are APIs on `CHIPDeviceController` that end up calling into `SetUpCodePairer` and then that calls back into `CHIPDeviceController`.  With this change there can be more round-trips of that, but no fundamental difference.

If the APIs on `CHIPDeviceController` that call into `SetUpCodePairer` are never used, then there should be no difference in behavior, as long as we ensure that `TryNextRendezvousParameters` returns false in that situation.  Which I will make sure happens.

Reply:
One potentially incorrect case is if (for some strange reason), PairDevice is called with the setup code (BLE and IP). Device is discovered on IP and commissioned. Second PairDevice is called NOT through the setupcode pairer for a device on IP. If the PASE connection on #2 fails, I think TryNextRendezvousParameters will still return true because of the pending BLE from the first PairDevice call. 

That can be fixed, I believe, by calling ResetDiscoveryState when the pase connection is established.

Reply:
I do share Tennessee's concerns though - the more tightly we could these, the harder it gets to disentangle them at a later date.

Reply:
Approving assuming a follow-up comes to reduce the risk of SetUpCodePairer complex state hooks interfering with DeviceCommissioner users that don't want to rely on SetupCodePairer for their discovery attempts.